src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,153 @@
+        {
+            throw new NotImplementedException();
+        }
+
+        private ValueTask<Stream> HandleSharedMemoryRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private async Task<Stream> HandleTcpRequest(TcpTransportCreationParams request, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            var ipAddresses = await Dns.GetHostAddressesAsync(request.Hostname, ct).ConfigureAwait(false);
+            if (ipAddresses.Length == 0)
+            {
+                throw new Exception("Hostname did not resolve");
+            }
+
+            // If there is an IP version preference, apply it
+            switch (request.IpAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV6);
+                    break;
+
+                case SqlConnectionIPAddressPreference.UsePlatformDefault:
+                default:
+                    // Not sorting necessary
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketExceptions = new List<Exception>();
+            foreach (var ipAddress in ipAddresses)
+            {
+                var ipEndpoint = new IPEndPoint(ipAddress, request.Port);
+                try
+                {
+                    socket = await OpenSocketAsync(ipEndpoint, ct).ConfigureAwait(false);
+                }
+                catch(Exception e)
+                {
+                    socketExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                throw new AggregateException(socketExceptions);
+            }
+
+            // Create the stream for the socket
+            return new NetworkStream(socket);
+        }
+
+        private async Task<Socket> OpenSocketAsync(IPEndPoint ipEndpoint, CancellationToken ct)
+        {
+            var socket = new Socket(ipEndpoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp) { Blocking = false, };
+
+            // Enable keep-alive
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, 1);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, 30);
+
+            // Open the socket
+            try
+            {
+                await socket.ConnectAsync(ipEndpoint, ct).ConfigureAwait(false);
```

Comment:
This overload isn't supported in .NET Framework - in that situation, you might need to spool up a new Task with the synchronous Connect and wait on it with the CancellationToken. It's not ideal.

Reply:
Ah, I see why you mentioned OpenSocketAsync might be sync over async. I'm wondering if using the `TaskFactory.FromAsync(BeginConnect, EndConnect)` would be a better option than `Task.Run(() => Connect())`. Obviously this would only be used if we're in netfx.

Reply:
I've looked through `Socket.ConnectAsync` and `TaskFactory`. Unfortunately the default CancellationToken passed in the TaskFactory constructor doesn't flow through to the `FromAsync` methods, so that won't work as we expect.

The only way to cancel the async connection attempt might simply be to register a handler against the CancellationToken which disposes of the socket, similar to the below:

```csharp
ct.Register((state) => ((Socket)state!).Dispose(), socket);
```

Reply:
Ok, I think handling cancellation and timeouts will be held for another PR. But I will definitely keep this in mind when addressing that.